proposals/0010-resource-element-type-validation.md

```diff
@@ -0,0 +1,223 @@
+valid RET, there should be infrastructure to reject this RET and emit a message 
+explaining why it was rejected as an RET.
+
+## Motivation
+Currently, there is an allow list of valid RETs. It must be modified with respect 
+to this spec. Anything that is not a valid RET will be rejected. The allow list isn't
+broad enough, because there is no case where the RET is user-defined for a raw buffer.
+Ideally, a user should be able to determine how any user-defined structure is invalid 
+as an RET. Some system should be in place to more completely enforce the rules for 
+valid and invalid RETs, as well as provide useful information on why they are invalid.
+
+For example, when targeting DXIL IR, `RWBuffer<double4> b : register(u4);` will emit
+an error in DXC, but will not in clang-dxc, despite the fact that `double4` is an 
+invalid RET for typed buffers.
+
+## Proposed solution
+
+The proposed solution is to modify the declaration of each resource declared in 
+`clang\lib\Sema\HLSLExternalSemaSource.cpp` and insert into each representative
+AST node a concept. The AST node will be created as if the C++20 `concept` keyword
+was parsed and applied to the declaration. The concept will be used to validate the
+given RET, and will emit errors when the given RET is invalid. Although concepts are
+not currently supported in HLSL, we expect support to be added at some point in the
+future. Meanwhile, because LLVM does support concepts, we can make use of
+them when constructing the AST in Sema.
+
+A new built-in, `__builtin_hlsl_is_line_vector_layout_compatible`, will be 
+added in order to express the extra typed buffer constraint. This builtin
+will be added to each AST node that requires that constraint. The builtin is 
+described below. Standard clang diagnostics for unsatisfied constraints will be 
+used to report any invalid element types. Concepts required will differ depending
+on whether the resource is a typed buffer or a raw buffer. Until concepts are 
+formally supported by HLSL, the concepts and constraints will be expressed 
+directly in the AST via the HLSL external sema source.
+
+## Detailed design
+
+In `clang\lib\Sema\HLSLExternalSemaSource.cpp`, `RWBuffer` is defined, along with 
+`RasterizerOrderedBuffer` and `StructuredBuffer`. It is at this point that the 
+concept would be incorporated into these resource declarations. A concept representing
+the relevant constraints will be applied to each resource declaration. If a concept
+is not true for the given RET, a corresponding error message will be emitted.
+
+The list of builtins to be used as type traits that will be available for
+concept definition are described below:
+| type trait | Description|
+|-|-|
+| `__is_intangible` | An RET should be an arithmetic type, bool, enum, or a vector or matrix or UDT containing such types. This is equivalent to validating that the RET is not intangible. This will error when given an incomplete type. |

```

Comment:
The full name is `__builtin_hlsl_is_intangible` and I would probably use list it here with a negation `!__builtin_hlsl_is_intangible`, otherwise it does not match the description on the right.